### PR TITLE
fix: prevent duplicate `<link>` module preloads

### PR DESCRIPTION
**Pull request by s3812497**
fixes https://github.com/sveltejs/kit/issues/10358
fixes https://github.com/sveltejs/kit/issues/10379

~This PR disables [Vite's module preload](https://vitejs.dev/config/build-options.html#build-modulepreload) for JS files only (Vite still injects CSS) which adds the preloads on [startup](https://github.com/vitejs/vite/blob/1292ad08a06e2ae8c44307797ba533f64a602759/packages/vite/src/node/plugins/importAnalysisBuild.ts#L121-L128) in addition to the preloads already [added by Kit](https://github.com/sveltejs/kit/blob/b251b235b321b9c75236429cc3a1ac2415475775/packages/kit/src/runtime/server/page/render.js#L240). I'm not currently aware of any side-effects disabling this may have so this PR might be somewhat naive.~

This PR matches the Vite base path setting [optimisation](https://github.com/sveltejs/kit/pull/10287/files) with the [page rendering](https://github.com/sveltejs/kit/blob/926c22d94fe7884dc9bc020ce735a911f50c8509/packages/kit/src/runtime/server/page/render.js#L96C4-L108) base path setting, resulting in identical hrefs. Hence, Vite no longer duplicates asset links that are already in the `<head>`.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by s3812497** [comment]
> Crazy, it's failing a seemingly completely unrelated test, but consistently.

Apparently, disabling Vite's module preload causes client-side navigation to not load the subsequent page's CSS files. Hence, the test is failing because the required CSS rules are not applied.

**Comment by s3812497** [comment]
Instead of messing with Vite's module preload option, we might have to prefer absolute asset URLs by default.
[This part of page rendering](https://github.com/sveltejs/kit/blob/926c22d94fe7884dc9bc020ce735a911f50c8509/packages/kit/src/runtime/server/page/render.js#L96C4-L108) that prefers relative URLs conflicts with this [perf optimisation](https://github.com/sveltejs/kit/pull/10287/files) that prefers absolute URLs, causing the asset URLs to be different and Vite unavoidably duplicates them.

**Comment by dummdidumm** [comment]
To me it sounds like we should then revert that perf optimisation then for the time being.

**Comment by s3812497** [comment]
I've adjusted so that it is no longer the default, but the optimisation still works if the `relative` path setting is set to `false`.

File: packages/kit/src/exports/vite/index.js [hunk 1]
```diff
@@ -545,7 +545,8 @@ function kit({ svelte_config }) {
 				// E.g. Vite generates `new URL('/asset.png', import.meta).href` for a relative path vs just '/asset.png'.
 				// That's larger and takes longer to run and also causes an HTML diff between SSR and client
 				// causing us to do a more expensive hydration check.
-				const client_base = kit.paths.relative || kit.paths.assets ? './' : kit.paths.base || '/';
+				const client_base =
+					kit.paths.relative !== false || kit.paths.assets ? './' : kit.paths.base || '/';
```
Comment (benmccann) [inline packages/kit/src/exports/vite/index.js:549]: why does this need to be changed to `!== false`?

**Comment by benmccann** [comment]
> the optimisation still works if the relative path setting is set to false

won't that create duplicate module preload `<link>` elements in that case though?

it seems to me like the better way might be to update that section of code you referred to: https://github.com/sveltejs/kit/blob/926c22d94fe7884dc9bc020ce735a911f50c8509/packages/kit/src/runtime/server/page/render.js#L96C4-L108

**Comment by s3812497** [comment]
> > the optimisation still works if the relative path setting is set to false
> 
> won't that create duplicate module preload `<link>` elements in that case though?

It shouldn't as long as the base path settings match for both sections of code (either `./` or whatever `kit.config.base` is).

> it seems to me like the better way might be to update that section of code you referred to: [`926c22d`/packages/kit/src/runtime/server/page/render.js#L96C4-L108](https://github.com/sveltejs/kit/blob/926c22d94fe7884dc9bc020ce735a911f50c8509/packages/kit/src/runtime/server/page/render.js#L96C4-L108)

Yes, initially I was inclined to change it so that absolute paths was the default, but I'm not sure if that would cause problems for existing apps that relied on the default setting (relative paths)?
```diff
- if (paths.relative !== false && !state.prerendering?.fallback) {
+ if (paths.relative && !state.prerendering?.fallback) {
```
The default behaviour as mentioned in the docs:
> By default, if paths.assets is not external, SvelteKit will replace %sveltekit.assets% with a relative path and use relative paths to reference build artifacts

We would be changing the default to `false`.
> %sveltekit.assets% and references to build artifacts will always be root-relative paths, unless paths.assets is an external URL

https://kit.svelte.dev/docs/configuration#paths

**Review by dummdidumm — APPROVED** [review]
Agree with that reasoning 👍

**Pull request merged**